### DISCUSSION TITLE
enable M7/M8/M9 in Marlin for laser engraving

### DISCUSSION THREAD

**Discussion by armature81** [arkypita/LaserGRBL#1335] [General]
I tried to enable M7/M8/M9 in marlin 2.7.2 for coolant ( air assist)  when i use laser module in my 3d printer
when I do enable and configure for M7/M8 i get this error as below
can you please help me to solved

Arduino: 1.8.13 (Windows 10), Board: “Arduino Mega or Mega 2560, ATmega2560 (Mega 2560)”

Invalid library found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\ESP8266-SerialWebSocket-master: no headers files (.h) found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\ESP8266-SerialWebSocket-master

Invalid library found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\websocketserver: no headers files (.h) found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\websocketserver

In file included from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\hal.h:22:0,

             from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\hal.h:26,

             from sketch\src\inc/MarlinConfig.h:31,

             from sketch\src\MarlinCore.h:24,

             from sketch\src\MarlinCore.cpp:31:
sketch\src\MarlinCore.cpp: In function ‘void setup()’:

c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:87:35: error: ‘DIOCOOLANT_FLOOD_PIN_DDR’ was not declared in this scope

#define _SET_OUTPUT(IO) SBI(DIO ## IO ## _DDR, DIO ## IO ## _PIN)

                               ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:21: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:108:31: note: in expansion of macro ‘_SET_OUTPUT’

#define SET_OUTPUT(IO) _SET_OUTPUT(IO)

                           ^~~~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:35: note: in expansion of macro ‘SET_OUTPUT’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                               ^~~~~~~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:87:35: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

#define _SET_OUTPUT(IO) SBI(DIO ## IO ## _DDR, DIO ## IO ## _PIN)

                               ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:21: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:108:31: note: in expansion of macro ‘_SET_OUTPUT’

#define SET_OUTPUT(IO) _SET_OUTPUT(IO)

                           ^~~~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:35: note: in expansion of macro ‘SET_OUTPUT’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                               ^~~~~~~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:87:54: error: ‘DIOCOOLANT_FLOOD_PIN_PIN’ was not declared in this scope

#define _SET_OUTPUT(IO) SBI(DIO ## IO ## _DDR, DIO ## IO ## _PIN)

                                                  ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:33: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                             ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:108:31: note: in expansion of macro ‘_SET_OUTPUT’

#define SET_OUTPUT(IO) _SET_OUTPUT(IO)

                           ^~~~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:35: note: in expansion of macro ‘SET_OUTPUT’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                               ^~~~~~~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:87:54: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

#define _SET_OUTPUT(IO) SBI(DIO ## IO ## _DDR, DIO ## IO ## _PIN)

                                                  ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:33: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                             ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:108:31: note: in expansion of macro ‘_SET_OUTPUT’

#define SET_OUTPUT(IO) _SET_OUTPUT(IO)

                           ^~~~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:35: note: in expansion of macro ‘SET_OUTPUT’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                               ^~~~~~~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
In file included from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\hal.h:24:0,

             from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\hal.h:26,

             from sketch\src\inc/MarlinConfig.h:31,

             from sketch\src\MarlinCore.h:24,

             from sketch\src\MarlinCore.cpp:31:
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:41: error: ‘DIOCOOLANT_FLOOD_PIN_RPORT’ was not declared in this scope

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                     ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:41: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                     ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
In file included from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\hal.h:22:0,

             from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\hal.h:26,

             from sketch\src\inc/MarlinConfig.h:31,

             from sketch\src\MarlinCore.h:24,

             from sketch\src\MarlinCore.cpp:31:
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:72:14: error: ‘DIOCOOLANT_FLOOD_PIN_WPORT’ was not declared in this scope

if (V) SBI(DIO ## IO ## _WPORT, DIO ## IO ## _PIN); \

          ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:21: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:81: note: in expansion of macro ‘_WRITE_NC’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                             ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:72:14: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

if (V) SBI(DIO ## IO ## _WPORT, DIO ## IO ## _PIN); \

          ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:48:21: note: in definition of macro ‘SBI’

#define SBI(A,B) (A |= (1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:81: note: in expansion of macro ‘_WRITE_NC’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                             ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:73:14: error: ‘DIOCOOLANT_FLOOD_PIN_WPORT’ was not declared in this scope

else CBI(DIO ## IO ## _WPORT, DIO ## IO ## _PIN); \

          ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:52:21: note: in definition of macro ‘CBI’

#define CBI(A,B) (A &= ~(1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:81: note: in expansion of macro ‘_WRITE_NC’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                             ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:73:14: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

else CBI(DIO ## IO ## _WPORT, DIO ## IO ## _PIN); \

          ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\shared\marduino.h:52:21: note: in definition of macro ‘CBI’

#define CBI(A,B) (A &= ~(1 << (B)))

                 ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:81: note: in expansion of macro ‘_WRITE_NC’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                             ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
In file included from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\hal.h:24:0,

             from c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\hal.h:26,

             from sketch\src\inc/MarlinConfig.h:31,

             from sketch\src\MarlinCore.h:24,

             from sketch\src\MarlinCore.cpp:31:
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:77:23: error: ‘DIOCOOLANT_FLOOD_PIN_WPORT’ was not declared in this scope

uint8_t port_bits = DIO ## IO ## _WPORT; /* Get a mask from the current port bits */ \

                   ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:103: note: in expansion of macro ‘_WRITE_C’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                                                   ^~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:77:23: note: suggested alternative: ‘COOLANT_FLOOD_INVERT’

uint8_t port_bits = DIO ## IO ## _WPORT; /* Get a mask from the current port bits */ \

                   ^
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:103: note: in expansion of macro ‘_WRITE_C’

#define _WRITE(IO,V) do{ if (&(DIO ## IO ## _RPORT) < (uint8_t*)0x100) _WRITE_NC(IO,V); else _WRITE_C(IO,V); }while(0)

                                                                                                   ^~~~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:102:31: note: in expansion of macro ‘_WRITE’

#define WRITE(IO,V) _WRITE(IO,V)

                           ^~~~~~
c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:114:51: note: in expansion of macro ‘WRITE’

#define OUT_WRITE(IO,V) do{ SET_OUTPUT(IO); WRITE(IO,V); }while(0)

                                               ^~~~~
sketch\src\MarlinCore.cpp:1129:5: note: in expansion of macro ‘OUT_WRITE’

 OUT_WRITE(COOLANT_FLOOD_PIN, COOLANT_FLOOD_INVERT); // Init Flood Coolant OFF

 ^~~~~~~~~
exit status 1

Error compiling for board Arduino Mega or Mega 2560.

Invalid library found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\ESP8266-SerialWebSocket-master: no headers files (.h) found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\ESP8266-SerialWebSocket-master

Invalid library found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\websocketserver: no headers files (.h) found in C:\Users\armat\OneDrive\Documents\Arduino\libraries\websocketserver

This report would have more information with
“Show verbose output during compilation”
option enabled in File -> Preferences.

**Comment by gmmanonymus111** [comment]
`c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:73:14: error: ‘DIOCOOLANT_FLOOD_PIN_WPORT’ was not declared in this scope`
`c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:87:54: error: ‘DIOCOOLANT_FLOOD_PIN_PIN’ was not declared in this scope`
`c:\users\armat\appdata\local\temp\arduino_build_252812\sketch\src\hal\avr\fastio.h:82:41: error: ‘DIOCOOLANT_FLOOD_PIN_RPORT’ was not declared in this scope`
`...`
Looks like you are missing some include files, forgot to declare some variables, or your enviroment is not set up properly.

You should [check here](https://github.com/MarlinFirmware/Marlin#marlin-support) for Marlin help.
Please, next time use code tags, as the error message is really hard to read this way.
